// @openai/agents-core/dist/shims/mcp-server/browser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MCPServerSSE = exports.MCPServerStreamableHttp = exports.MCPServerStdio = void 0;
const mcp_1 = require("../../mcp.js");
class MCPServerStdio extends mcp_1.BaseMCPServerStdio {
    constructor(params) {
        super(params);
    }
    get name() {
        return 'MCPServerStdio';
    }
    connect() {
        throw new Error('Method not implemented.');
    }
    close() {
        throw new Error('Method not implemented.');
    }
    listTools() {
        throw new Error('Method not implemented.');
    }
    callTool(_toolName, _args, _meta) {
        throw new Error('Method not implemented.');
    }
    invalidateToolsCache() {
        throw new Error('Method not implemented.');
    }
}
exports.MCPServerStdio = MCPServerStdio;
class MCPServerStreamableHttp extends mcp_1.BaseMCPServerStreamableHttp {
    constructor(params) {
        super(params);
    }
    get name() {
        return 'MCPServerStdio';
    }
    connect() {
        throw new Error('Method not implemented.');
    }
    close() {
        throw new Error('Method not implemented.');
    }
    listTools() {
        throw new Error('Method not implemented.');
    }
    callTool(_toolName, _args, _meta) {
        throw new Error('Method not implemented.');
    }
    invalidateToolsCache() {
        throw new Error('Method not implemented.');
    }
}
exports.MCPServerStreamableHttp = MCPServerStreamableHttp;
class MCPServerSSE extends mcp_1.BaseMCPServerSSE {
    constructor(params) {
        super(params);
    }
    get name() {
        return 'MCPServerSSE';
    }
    connect() {
        throw new Error('Method not implemented.');
    }
    close() {
        throw new Error('Method not implemented.');
    }
    listTools() {
        throw new Error('Method not implemented.');
    }
    callTool(_toolName, _args, _meta) {
        throw new Error('Method not implemented.');
    }
    invalidateToolsCache() {
        throw new Error('Method not implemented.');
    }
}
exports.MCPServerSSE = MCPServerSSE;
