// @openai/agents-realtime/dist/openaiRealtimeWebRtc.js
"use strict";
/// <reference lib="dom" />
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAIRealtimeWebRTC = void 0;
const _shims_1 = require("@openai/agents-core/_shims");
const agents_core_1 = require("@openai/agents-core");
const logger_1 = __importDefault(require("./logger.js"));
const openaiRealtimeBase_1 = require("./openaiRealtimeBase.js");
const openaiRealtimeEvents_1 = require("./openaiRealtimeEvents.js");
const utils_1 = require("./utils.js");
/**
 * Transport layer that's handling the connection between the client and OpenAI's Realtime API
 * via WebRTC. While this transport layer is designed to be used within a RealtimeSession, it can
 * also be used standalone if you want to have a direct connection to the Realtime API.
 *
 * Unless you specify a `mediaStream` or `audioElement` option, the transport layer will
 * automatically configure the microphone and audio output to be used by the session.
 */
class OpenAIRealtimeWebRTC extends openaiRealtimeBase_1.OpenAIRealtimeBase {
    options;
    #url;
    #state = {
        status: 'disconnected',
        peerConnection: undefined,
        dataChannel: undefined,
        callId: undefined,
    };
    #useInsecureApiKey;
    #ongoingResponse = false;
    #muted = false;
    #connectPromise;
    #connectAttemptId = 0;
    constructor(options = {}) {
        if (typeof RTCPeerConnection === 'undefined') {
            throw new Error('WebRTC is not supported in this environment');
        }
        super(options);
        this.options = options;
        this.#url = options.baseUrl ?? `https://api.openai.com/v1/realtime/calls`;
        this.#useInsecureApiKey = options.useInsecureApiKey ?? false;
    }
    /**
     * The current call ID of the WebRTC connection.
     */
    get callId() {
        return this.#state.callId;
    }
    /**
     * The current status of the WebRTC connection.
     */
    get status() {
        return this.#state.status;
    }
    /**
     * The current connection state of the WebRTC connection including the peer connection and data
     * channel.
     */
    get connectionState() {
        return this.#state;
    }
    /**
     * Whether the session is muted.
     */
    get muted() {
        return this.#muted;
    }
    /**
     * Connect to the Realtime API. This will establish the connection to the OpenAI Realtime API
     * via WebRTC.
     *
     * If you are using a browser, the transport layer will also automatically configure the
     * microphone and audio output to be used by the session.
     *
     * @param options - The options for the connection.
     */
    async connect(options) {
        if (this.#state.status === 'connected') {
            return;
        }
        if (this.#state.status === 'connecting') {
            if (this.#connectPromise) {
                return this.#connectPromise;
            }
            logger_1.default.warn('Realtime connection already in progress but no promise found');
            return;
        }
        const model = options.model ?? this.currentModel;
        this.currentModel = model;
        const baseUrl = options.url ?? this.#url;
        const apiKey = await this._getApiKey(options);
        const isClientKey = typeof apiKey === 'string' && apiKey.startsWith('ek_');
        if ((0, _shims_1.isBrowserEnvironment)() && !this.#useInsecureApiKey && !isClientKey) {
            throw new agents_core_1.UserError('Using the WebRTC connection in a browser environment requires an ephemeral client key. If you need to use a regular API key, use the WebSocket transport or set the `useInsecureApiKey` option to true.');
        }
        const attemptId = ++this.#connectAttemptId;
        // eslint-disable-next-line no-async-promise-executor
        this.#connectPromise = new Promise(async (resolve, reject) => {
            try {
                const userSessionConfig = {
                    ...(options.initialSessionConfig || {}),
                    model: this.currentModel,
                };
                const connectionUrl = new URL(baseUrl);
                let peerConnection = new RTCPeerConnection();
                const dataChannel = peerConnection.createDataChannel('oai-events');
                let callId = undefined;
                const attachConnectionStateHandler = (connection) => {
                    connection.onconnectionstatechange = () => {
                        switch (connection.connectionState) {
                            case 'disconnected':
                            case 'failed':
                            case 'closed':
                                this.close();
                                break;
                            // 'connected' state is handled by dataChannel.onopen. So we don't need to handle it here.
                            // 'new' and 'connecting' are intermediate states and do not require action here.
                        }
                    };
                };
                attachConnectionStateHandler(peerConnection);
                this.#state = {
                    status: 'connecting',
                    peerConnection,
                    dataChannel,
                    callId,
                };
                this.emit('connection_change', this.#state.status);
                dataChannel.addEventListener('open', () => {
                    this.#state = {
                        status: 'connecting',
                        peerConnection,
                        dataChannel,
                        callId,
                    };
                    // Wait for session.updated acknowledgement before resolving connect().
                    // Without this, audio can flow to the server before config (instructions,
                    // tools, modalities) is applied, causing the server to use defaults.
                    let resolved = false;
                    // eslint-disable-next-line prefer-const -- declared before finish() to avoid TDZ if a callback fires synchronously
                    let timeoutId;
                    const finish = () => {
                        if (resolved)
                            return;
                        resolved = true;
                        if (timeoutId !== undefined)
                            clearTimeout(timeoutId);
                        dataChannel.removeEventListener('message', onConfigAck);
                        dataChannel.removeEventListener('close', onClose);
                        // Reject if the transport was closed/errored while waiting,
                        // the dataChannel is no longer open, or a different connection
                        // attempt is now active (stale timeout from an earlier connect).
                        if (this.#state.status !== 'connecting' ||
                            this.#state.dataChannel !== dataChannel ||
                            dataChannel.readyState !== 'open') {
                            // Transition to disconnected if this attempt is still the
                            // active one so that callers can retry connect() without
                            // needing to call close() first.
                            if (this.#state.dataChannel === dataChannel) {
                                this.close();
                            }
                            reject(new Error('Connection closed before session config was acknowledged'));
                            return;
                        }
                        this.#state = {
                            status: 'connected',
                            peerConnection,
                            dataChannel,
                            callId,
                        };
                        this.emit('connection_change', this.#state.status);
                        this._onOpen();
                        resolve();
                    };
                    const onConfigAck = (ackEvent) => {
                        const parsed = JSON.parse(ackEvent.data);
                        if (parsed.type === 'session.updated') {
                            finish();
                        }
                    };
                    const onClose = () => {
                        finish();
                    };
                    timeoutId = setTimeout(() => {
                        if (!resolved) {
                            logger_1.default.warn('Timed out waiting for session.updated ack — resolving connect() anyway');
                            finish();
                        }
                    }, 5000);
                    dataChannel.addEventListener('message', onConfigAck);
                    dataChannel.addEventListener('close', onClose);
                    // Register the general message handler AFTER onConfigAck so that
                    // finish() resolves connect() before _onMessage emits the
                    // session.updated event to external listeners.
                    dataChannel.addEventListener('message', (event) => {
                        this._onMessage(event);
                        const { data: parsed, isGeneric } = (0, openaiRealtimeEvents_1.parseRealtimeEvent)(event);
                        if (!parsed || isGeneric) {
                            return;
                        }
                        if (parsed.type === 'response.created') {
                            this.#ongoingResponse = true;
                        }
                        else if (parsed.type === 'response.done') {
                            this.#ongoingResponse = false;
                        }
                        if (parsed.type === 'session.created') {
                            this._tracingConfig = parsed.session.tracing;
                            // Trying to turn on tracing after the session is created
                            const tracingConfig = typeof userSessionConfig.tracing === 'undefined'
                                ? 'auto'
                                : userSessionConfig.tracing;
                            this._updateTracingConfig(tracingConfig);
                        }
                    });
                    this.updateSessionConfig(userSessionConfig);
                });
                dataChannel.addEventListener('error', (event) => {
                    this.close();
                    this._onError(event);
                    reject(event);
                });
                // set up audio playback
                const audioElement = this.options.audioElement ?? document.createElement('audio');
                audioElement.autoplay = true;
                peerConnection.ontrack = (event) => {
                    audioElement.srcObject = event.streams[0];
                };
                // get microphone stream
                const stream = this.options.mediaStream ??
                    (await navigator.mediaDevices.getUserMedia({
                        audio: true,
                    }));
                peerConnection.addTrack(stream.getAudioTracks()[0]);
                if (this.options.changePeerConnection) {
                    const originalPeerConnection = peerConnection;
                    peerConnection =
                        await this.options.changePeerConnection(peerConnection);
                    if (originalPeerConnection !== peerConnection) {
                        originalPeerConnection.onconnectionstatechange = null;
                    }
                    attachConnectionStateHandler(peerConnection);
                    this.#state = { ...this.#state, peerConnection };
                }
                const offer = await peerConnection.createOffer();
                await peerConnection.setLocalDescription(offer);
                if (!offer.sdp) {
                    throw new Error('Failed to create offer');
                }
                const sdpResponse = await fetch(connectionUrl, {
                    method: 'POST',
                    body: offer.sdp,
                    headers: {
                        'Content-Type': 'application/sdp',
                        Authorization: `Bearer ${apiKey}`,
                        'X-OpenAI-Agents-SDK': utils_1.HEADERS['X-OpenAI-Agents-SDK'],
                    },
                });
                callId = sdpResponse.headers?.get('Location')?.split('/').pop();
                this.#state = { ...this.#state, callId };
                const answer = {
                    type: 'answer',
                    sdp: await sdpResponse.text(),
                };
                await peerConnection.setRemoteDescription(answer);
            }
            catch (error) {
                this.close();
                this._onError(error);
                reject(error);
            }
        }).finally(() => {
            // Only clear if this is still the active connection attempt.
            // A newer connect() may have already replaced #connectPromise.
            if (this.#connectAttemptId === attemptId) {
                this.#connectPromise = undefined;
            }
        });
        return this.#connectPromise;
    }
    /**
     * Send an event to the Realtime API. This will stringify the event and send it directly to the
     * API. This can be used if you want to take control over the connection and send events manually.
     *
     * @param event - The event to send.
     */
    sendEvent(event) {
        if (!this.#state.dataChannel ||
            this.#state.dataChannel.readyState !== 'open') {
            throw new Error('WebRTC data channel is not connected. Make sure you call `connect()` before sending events.');
        }
        this.#state.dataChannel.send(JSON.stringify(event));
    }
    /**
     * Mute or unmute the session.
     * @param muted - Whether to mute the session.
     */
    mute(muted) {
        this.#muted = muted;
        if (this.#state.peerConnection) {
            const peerConnection = this.#state.peerConnection;
            peerConnection.getSenders().forEach((sender) => {
                if (sender.track) {
                    sender.track.enabled = !muted;
                }
            });
        }
    }
    _afterAudioDoneEvent() {
        this.#ongoingResponse = false;
    }
    /**
     * Close the connection to the Realtime API and disconnects the underlying WebRTC connection.
     */
    close() {
        if (this.#state.dataChannel) {
            this.#state.dataChannel.close();
        }
        if (this.#state.peerConnection) {
            const peerConnection = this.#state.peerConnection;
            peerConnection.onconnectionstatechange = null;
            peerConnection.getSenders().forEach((sender) => {
                sender.track?.stop();
            });
            peerConnection.close();
        }
        if (this.#state.status !== 'disconnected') {
            this.#state = {
                status: 'disconnected',
                peerConnection: undefined,
                dataChannel: undefined,
                callId: undefined,
            };
            this.emit('connection_change', this.#state.status);
            this._onClose();
        }
    }
    /**
     * Interrupt the current response if one is ongoing and clear the audio buffer so that the agent
     * stops talking.
     */
    interrupt() {
        if (this.#ongoingResponse) {
            this.sendEvent({
                type: 'response.cancel',
            });
            this.#ongoingResponse = false;
        }
        this.sendEvent({
            type: 'output_audio_buffer.clear',
        });
    }
}
exports.OpenAIRealtimeWebRTC = OpenAIRealtimeWebRTC;
