// @openai/agents-core/dist/utils/zodJsonSchemaCompat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasJsonSchemaObjectShape = hasJsonSchemaObjectShape;
exports.zodJsonSchemaCompat = zodJsonSchemaCompat;
exports.mergeJsonSchemaDescriptions = mergeJsonSchemaDescriptions;
const zodCompat_1 = require("./zodCompat.js");
const JSON_SCHEMA_DRAFT_07 = 'http://json-schema.org/draft-07/schema#';
const OPTIONAL_WRAPPERS = new Set(['optional']);
const DECORATOR_WRAPPERS = new Set([
    'brand',
    'branded',
    'catch',
    'default',
    'effects',
    'pipeline',
    'pipe',
    'prefault',
    'readonly',
    'refinement',
    'transform',
]);
// Primitive leaf nodes map 1:1 to JSON Schema types; everything else is handled
// by the specialised builders further down.
const SIMPLE_TYPE_MAPPING = {
    string: { type: 'string' },
    number: { type: 'number' },
    bigint: { type: 'integer' },
    boolean: { type: 'boolean' },
    date: { type: 'string', format: 'date-time' },
};
function hasJsonSchemaObjectShape(value) {
    return (typeof value === 'object' &&
        value !== null &&
        value.type === 'object' &&
        'properties' in value &&
        'additionalProperties' in value);
}
function zodJsonSchemaCompat(input) {
    // Attempt to build an object schema from Zod's internal shape. If we cannot
    // understand the structure we return undefined, letting callers raise a
    // descriptive error instead of emitting an invalid schema.
    const schema = buildObjectSchema(input);
    if (!schema) {
        return undefined;
    }
    if (!Array.isArray(schema.required)) {
        schema.required = [];
    }
    if (typeof schema.additionalProperties === 'undefined') {
        schema.additionalProperties = false;
    }
    if (typeof schema.$schema !== 'string') {
        schema.$schema = JSON_SCHEMA_DRAFT_07;
    }
    return schema;
}
function mergeJsonSchemaDescriptions(target, source) {
    if (typeof target !== 'object' ||
        target === null ||
        typeof source !== 'object' ||
        source === null) {
        return;
    }
    const targetRecord = target;
    const sourceRecord = source;
    if (typeof sourceRecord.description === 'string' &&
        sourceRecord.description.trim() &&
        !('description' in targetRecord)) {
        targetRecord.description = sourceRecord.description;
    }
    if (targetRecord.type === 'object' &&
        sourceRecord.type === 'object' &&
        typeof targetRecord.properties === 'object' &&
        targetRecord.properties !== null &&
        typeof sourceRecord.properties === 'object' &&
        sourceRecord.properties !== null) {
        for (const [key, value] of Object.entries(sourceRecord.properties)) {
            const targetValue = targetRecord.properties[key];
            if (targetValue) {
                mergeJsonSchemaDescriptions(targetValue, value);
            }
        }
    }
    if (targetRecord.type === 'array' &&
        sourceRecord.type === 'array' &&
        'items' in targetRecord &&
        'items' in sourceRecord) {
        const targetItems = targetRecord.items;
        const sourceItems = sourceRecord.items;
        if (Array.isArray(targetItems) && Array.isArray(sourceItems)) {
            const limit = Math.min(targetItems.length, sourceItems.length);
            for (let index = 0; index < limit; index += 1) {
                mergeJsonSchemaDescriptions(targetItems[index], sourceItems[index]);
            }
        }
        else if (typeof targetItems === 'object' &&
            targetItems !== null &&
            typeof sourceItems === 'object' &&
            sourceItems !== null) {
            mergeJsonSchemaDescriptions(targetItems, sourceItems);
        }
    }
    for (const keyword of ['anyOf', 'allOf', 'oneOf']) {
        const targetArray = targetRecord[keyword];
        const sourceArray = sourceRecord[keyword];
        if (Array.isArray(targetArray) && Array.isArray(sourceArray)) {
            const limit = Math.min(targetArray.length, sourceArray.length);
            for (let index = 0; index < limit; index += 1) {
                mergeJsonSchemaDescriptions(targetArray[index], sourceArray[index]);
            }
        }
    }
}
function buildObjectSchema(value) {
    const shape = readShape(value);
    if (!shape) {
        return undefined;
    }
    const properties = {};
    const required = [];
    for (const [key, field] of Object.entries(shape)) {
        const { schema, optional } = convertProperty(field);
        if (!schema) {
            return undefined;
        }
        const description = readZodDescription(field);
        if (description &&
            typeof schema === 'object' &&
            schema !== null &&
            !('description' in schema)) {
            schema.description = description;
        }
        properties[key] = schema;
        if (!optional) {
            required.push(key);
        }
    }
    const schema = {
        type: 'object',
        properties,
        required,
        additionalProperties: false,
    };
    const description = readZodDescription(value);
    if (description) {
        schema.description = description;
    }
    return schema;
}
function convertProperty(value) {
    // Remove wrapper decorators (brand, transform, etc.) before attempting to
    // classify the node, tracking whether we crossed an `optional` boundary so we
    // can populate the `required` array later.
    let current = unwrapDecorators(value);
    let optional = false;
    while (OPTIONAL_WRAPPERS.has((0, zodCompat_1.readZodType)(current) ?? '')) {
        optional = true;
        const def = (0, zodCompat_1.readZodDefinition)(current);
        const next = unwrapDecorators(def?.innerType);
        if (!next || next === current) {
            break;
        }
        current = next;
    }
    return { schema: convertSchema(current), optional };
}
function convertSchema(value) {
    if (value === undefined) {
        return undefined;
    }
    const unwrapped = unwrapDecorators(value);
    const type = (0, zodCompat_1.readZodType)(unwrapped);
    const def = (0, zodCompat_1.readZodDefinition)(unwrapped);
    if (!type) {
        return undefined;
    }
    if (type in SIMPLE_TYPE_MAPPING) {
        return { ...SIMPLE_TYPE_MAPPING[type] };
    }
    switch (type) {
        case 'object':
            return buildObjectSchema(unwrapped);
        case 'array':
            return buildArraySchema(def);
        case 'tuple':
            return buildTupleSchema(def);
        case 'union':
            return buildUnionSchema(def);
        case 'intersection':
            return buildIntersectionSchema(def);
        case 'literal':
            return buildLiteral(def);
        case 'enum':
        case 'nativeenum':
            return buildEnum(def);
        case 'record':
            return buildRecordSchema(def);
        case 'map':
            return buildMapSchema(def);
        case 'set':
            return buildSetSchema(def);
        case 'nullable':
            return buildNullableSchema(def);
        default:
            return undefined;
    }
}
// --- JSON Schema builders -------------------------------------------------
function buildArraySchema(def) {
    const items = convertSchema(extractFirst(def, 'element', 'items', 'type'));
    return items ? { type: 'array', items } : undefined;
}
function buildTupleSchema(def) {
    const items = coerceArray(def?.items)
        .map((item) => convertSchema(item))
        .filter(Boolean);
    if (!items.length) {
        return undefined;
    }
    const schema = {
        type: 'array',
        items,
        minItems: items.length,
    };
    if (!def?.rest) {
        schema.maxItems = items.length;
    }
    return schema;
}
function buildUnionSchema(def) {
    const options = coerceArray(def?.options ?? def?.schemas)
        .map((option) => convertSchema(option))
        .filter(Boolean);
    return options.length ? { anyOf: options } : undefined;
}
function buildIntersectionSchema(def) {
    const left = convertSchema(def?.left);
    const right = convertSchema(def?.right);
    return left && right ? { allOf: [left, right] } : undefined;
}
function buildRecordSchema(def) {
    const valueSchema = convertSchema(def?.valueType ?? def?.values);
    return valueSchema
        ? { type: 'object', additionalProperties: valueSchema }
        : undefined;
}
function buildMapSchema(def) {
    const valueSchema = convertSchema(def?.valueType ?? def?.values);
    return valueSchema ? { type: 'array', items: valueSchema } : undefined;
}
function buildSetSchema(def) {
    const valueSchema = convertSchema(def?.valueType);
    return valueSchema
        ? { type: 'array', items: valueSchema, uniqueItems: true }
        : undefined;
}
function buildNullableSchema(def) {
    const inner = convertSchema(def?.innerType ?? def?.type);
    return inner ? { anyOf: [inner, { type: 'null' }] } : undefined;
}
function unwrapDecorators(value) {
    let current = value;
    while (DECORATOR_WRAPPERS.has((0, zodCompat_1.readZodType)(current) ?? '')) {
        const def = (0, zodCompat_1.readZodDefinition)(current);
        const next = def?.innerType ??
            def?.schema ??
            def?.base ??
            def?.type ??
            def?.wrapped ??
            def?.underlying;
        if (!next || next === current) {
            return current;
        }
        current = next;
    }
    return current;
}
function readZodDescription(value) {
    if (typeof value === 'object' && value !== null) {
        const direct = value.description;
        if (typeof direct === 'string' && direct.trim()) {
            return direct;
        }
    }
    let current = value;
    const visited = new Set();
    while (current && typeof current === 'object' && !visited.has(current)) {
        visited.add(current);
        const def = (0, zodCompat_1.readZodDefinition)(current);
        if (typeof def?.description === 'string' && def.description.trim()) {
            return def.description;
        }
        const next = def?.innerType ??
            def?.schema ??
            def?.base ??
            def?.type ??
            def?.wrapped ??
            def?.underlying;
        if (!next || next === current) {
            break;
        }
        current = next;
    }
    return undefined;
}
function extractFirst(def, ...keys) {
    if (!def) {
        return undefined;
    }
    for (const key of keys) {
        if (key in def && def[key] !== undefined) {
            return def[key];
        }
    }
    return undefined;
}
function coerceArray(value) {
    if (Array.isArray(value)) {
        return value;
    }
    return value === undefined ? [] : [value];
}
function buildLiteral(def) {
    if (!def) {
        return undefined;
    }
    const literal = extractFirst(def, 'value', 'literal');
    if (literal === undefined) {
        return undefined;
    }
    return {
        const: literal,
        type: literal === null ? 'null' : typeof literal,
    };
}
function buildEnum(def) {
    if (!def) {
        return undefined;
    }
    let values;
    if (Array.isArray(def.values)) {
        // Zod v3 z.enum() — values is a string array
        values = def.values;
    }
    else if (def.entries && typeof def.entries === 'object') {
        // Zod v4 z.enum() / z.nativeEnum() — entries is an object
        values = resolveNativeEnumValues(def.entries);
    }
    else if (Array.isArray(def.options)) {
        values = def.options;
    }
    else if (def.values && typeof def.values === 'object') {
        // Zod v3 z.nativeEnum() — values is the TS enum object
        values = resolveNativeEnumValues(def.values);
    }
    else if (def.enum && typeof def.enum === 'object') {
        // Fallback for alternative Zod internals
        values = resolveNativeEnumValues(def.enum);
    }
    if (!values || !values.length) {
        return undefined;
    }
    // @see https://github.com/StefanTerdell/zod-to-json-schema/blob/master/src/parsers/enum.ts
    const parsedTypes = Array.from(new Set(values.map((v) => typeof v)));
    const type = parsedTypes.length === 1
        ? parsedTypes[0] === 'string'
            ? 'string'
            : 'number'
        : ['string', 'number'];
    return { type, enum: values };
}
/**
 * Filter TypeScript's reverse-mapping keys from numeric native enums.
 * e.g. `enum E { A = 0 }` compiles to `{ A: 0, "0": "A" }` — keep only forward mappings.
 * @see https://github.com/StefanTerdell/zod-to-json-schema/blob/master/src/parsers/nativeEnum.ts#L12-L15
 */
function resolveNativeEnumValues(enumObj) {
    const actualKeys = Object.keys(enumObj).filter((key) => typeof enumObj[enumObj[key]] !== 'number');
    return actualKeys.map((key) => enumObj[key]);
}
function readShape(input) {
    if (typeof input !== 'object' || input === null) {
        return undefined;
    }
    const candidate = input;
    if (candidate.shape && typeof candidate.shape === 'object') {
        return candidate.shape;
    }
    if (typeof candidate.shape === 'function') {
        try {
            return candidate.shape();
        }
        catch (_error) {
            return undefined;
        }
    }
    const def = (0, zodCompat_1.readZodDefinition)(candidate);
    const shape = def?.shape;
    if (shape && typeof shape === 'object') {
        return shape;
    }
    if (typeof shape === 'function') {
        try {
            return shape();
        }
        catch (_error) {
            return undefined;
        }
    }
    return undefined;
}
