// @openai/agents-core/dist/agent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Agent = void 0;
const lifecycle_1 = require("./lifecycle.js");
const mcp_1 = require("./mcp.js");
const defaultModel_1 = require("./defaultModel.js");
const runContext_1 = require("./runContext.js");
const tool_1 = require("./tool.js");
const handoff_1 = require("./handoff.js");
const run_1 = require("./run.js");
const runState_1 = require("./runState.js");
const tools_1 = require("./utils/tools.js");
const messages_1 = require("./utils/messages.js");
const typeGuards_1 = require("./utils/typeGuards.js");
const abortSignals_1 = require("./utils/abortSignals.js");
const errors_1 = require("./errors.js");
const logger_1 = __importDefault(require("./logger.js"));
const agentToolInput_1 = require("./agentToolInput.js");
const agentToolRunConfig_1 = require("./agentToolRunConfig.js");
const agentToolRunResults_1 = require("./agentToolRunResults.js");
const agentToolSourceRegistry_1 = require("./agentToolSourceRegistry.js");
/**
 * The class representing an AI agent configured with instructions, tools, guardrails, handoffs and more.
 *
 * We strongly recommend passing `instructions`, which is the "system prompt" for the agent. In
 * addition, you can pass `handoffDescription`, which is a human-readable description of the
 * agent, used when the agent is used inside tools/handoffs.
 *
 * Agents are generic on the context type. The context is a (mutable) object you create. It is
 * passed to tool functions, handoffs, guardrails, etc.
 */
class Agent extends lifecycle_1.AgentHooks {
    /**
     * Create an Agent with handoffs and automatically infer the union type for TOutput from the handoff agents' output types.
     */
    static create(config) {
        return new Agent({
            ...config,
            handoffs: config.handoffs,
            outputType: config.outputType,
            handoffOutputTypeWarningEnabled: false,
        });
    }
    static DEFAULT_MODEL_PLACEHOLDER = '';
    name;
    instructions;
    prompt;
    handoffDescription;
    handoffs;
    model;
    modelSettings;
    tools;
    mcpServers;
    inputGuardrails;
    outputGuardrails;
    outputType = 'text';
    toolUseBehavior;
    resetToolChoice;
    _toolsExplicitlyConfigured;
    constructor(config) {
        super();
        if (typeof config.name !== 'string' || config.name.trim() === '') {
            throw new errors_1.UserError('Agent must have a name.');
        }
        this.name = config.name;
        this.instructions = config.instructions ?? Agent.DEFAULT_MODEL_PLACEHOLDER;
        this.prompt = config.prompt;
        this.handoffDescription = config.handoffDescription ?? '';
        this.handoffs = config.handoffs ?? [];
        this.model = config.model ?? '';
        this.modelSettings = config.modelSettings ?? (0, defaultModel_1.getDefaultModelSettings)();
        this.tools = config.tools ?? [];
        this._toolsExplicitlyConfigured = config.tools !== undefined;
        this.mcpServers = config.mcpServers ?? [];
        this.inputGuardrails = config.inputGuardrails ?? [];
        this.outputGuardrails = config.outputGuardrails ?? [];
        if (config.outputType) {
            this.outputType = config.outputType;
        }
        this.toolUseBehavior = config.toolUseBehavior ?? 'run_llm_again';
        this.resetToolChoice = config.resetToolChoice ?? true;
        if (
        // The user sets a non-default model
        config.model !== undefined &&
            // The default model is gpt-5
            (0, defaultModel_1.isGpt5Default)() &&
            // However, the specified model is not a gpt-5 model
            (typeof config.model !== 'string' ||
                !(0, defaultModel_1.gpt5ReasoningSettingsRequired)(config.model)) &&
            // The model settings are not customized for the specified model
            config.modelSettings === undefined) {
            // In this scenario, we should use a generic model settings
            // because non-gpt-5 models are not compatible with the default gpt-5 model settings.
            // This is a best-effort attempt to make the agent work with non-gpt-5 models.
            this.modelSettings = {};
        }
        // --- Runtime warning for handoff output type compatibility ---
        if (config.handoffOutputTypeWarningEnabled === undefined ||
            config.handoffOutputTypeWarningEnabled) {
            if (this.handoffs && this.outputType) {
                const outputTypes = new Set([JSON.stringify(this.outputType)]);
                for (const h of this.handoffs) {
                    if ('outputType' in h && h.outputType) {
                        outputTypes.add(JSON.stringify(h.outputType));
                    }
                    else if ('agent' in h && h.agent.outputType) {
                        outputTypes.add(JSON.stringify(h.agent.outputType));
                    }
                }
                if (outputTypes.size > 1) {
                    logger_1.default.warn(`[Agent] Warning: Handoff agents have different output types: ${Array.from(outputTypes).join(', ')}. You can make it type-safe by using Agent.create({ ... }) method instead.`);
                }
            }
        }
    }
    /**
     * Output schema name.
     */
    get outputSchemaName() {
        if (this.outputType === 'text') {
            return 'text';
        }
        else if ((0, typeGuards_1.isZodObject)(this.outputType)) {
            return 'ZodOutput';
        }
        else if (typeof this.outputType === 'object') {
            return this.outputType.name;
        }
        throw new Error(`Unknown output type: ${this.outputType}`);
    }
    /**
     * Makes a copy of the agent, with the given arguments changed. For example, you could do:
     *
     * ```
     * const newAgent = agent.clone({ instructions: 'New instructions' })
     * ```
     *
     * @param config - A partial configuration to change.
     * @returns A new agent with the given changes.
     */
    clone(config) {
        return new Agent({
            ...this,
            ...config,
        });
    }
    asTool(options) {
        const { toolName, toolDescription, customOutputExtractor, needsApproval, parameters, inputBuilder, includeInputSchema, runConfig, runOptions, resumeState, isEnabled, onStream, } = options;
        // Event handlers are scoped to this agent tool instance and are not shared; we only support registration (no removal) to keep the API surface small.
        const eventHandlers = new Map();
        const emitEvent = async (event) => {
            // We intentionally keep only add semantics (no off) to reduce surface area; handlers are scoped to this agent tool instance.
            const specific = eventHandlers.get(event.event.type);
            const wildcard = eventHandlers.get('*');
            const candidates = [
                ...(onStream ? [onStream] : []),
                ...(specific ? Array.from(specific) : []),
                ...(wildcard ? Array.from(wildcard) : []),
            ];
            // Run all handlers in parallel so a slow onStream callback does not block on(...) handlers (and vice versa).
            await Promise.allSettled(candidates.map((handler) => Promise.resolve().then(() => handler(event))));
        };
        const resolvedToolName = toolName ?? (0, tools_1.toFunctionToolName)(this.name);
        const toolParameters = (parameters ??
            agentToolInput_1.AgentAsToolInputSchema);
        const hasCustomParameters = typeof parameters !== 'undefined';
        const includeSchema = includeInputSchema === true && hasCustomParameters;
        const shouldCaptureToolInput = hasCustomParameters ||
            includeSchema ||
            typeof inputBuilder === 'function';
        const schemaInfo = shouldCaptureToolInput
            ? (0, agentToolInput_1.buildStructuredInputSchemaInfo)(toolParameters, resolvedToolName, includeSchema)
            : undefined;
        const baseTool = (0, tool_1.tool)({
            name: resolvedToolName,
            description: toolDescription ?? '',
            parameters: toolParameters,
            strict: true,
            needsApproval: needsApproval,
            isEnabled,
            execute: async (params, context, details) => {
                const typedParams = params;
                const runContextBase = runOptions?.context instanceof runContext_1.RunContext
                    ? runOptions.context
                    : typeof runOptions?.context !== 'undefined'
                        ? new runContext_1.RunContext(runOptions.context)
                        : context instanceof runContext_1.RunContext
                            ? context
                            : typeof context !== 'undefined'
                                ? new runContext_1.RunContext(context)
                                : new runContext_1.RunContext();
                const agentToolInvocation = {
                    toolName: details?.toolCall?.name ?? baseTool.name,
                    toolCallId: details?.toolCall?.callId,
                    toolArguments: details?.toolCall?.arguments,
                };
                const shouldClearToolInput = !shouldCaptureToolInput &&
                    typeof runContextBase.toolInput !== 'undefined';
                const runContext = shouldCaptureToolInput &&
                    typeof runContextBase._forkWithToolInput === 'function'
                    ? runContextBase._forkWithToolInput(typedParams)
                    : shouldClearToolInput &&
                        typeof runContextBase._forkWithoutToolInput === 'function'
                        ? runContextBase._forkWithoutToolInput()
                        : runContextBase;
                const resolvedInput = await (0, agentToolInput_1.resolveAgentToolInput)({
                    params: typedParams,
                    schemaInfo,
                    inputBuilder,
                });
                if (typeof resolvedInput !== 'string' &&
                    !Array.isArray(resolvedInput)) {
                    throw new errors_1.ModelBehaviorError('Agent tool called with invalid input');
                }
                const inheritedRunConfig = (0, agentToolRunConfig_1.getInheritedAgentToolRunConfig)((0, agentToolRunConfig_1.getAgentToolParentRunConfigFromDetails)(details), runConfig);
                const nestedRunConfig = (0, agentToolRunConfig_1.mergeAgentToolRunConfig)(inheritedRunConfig, runConfig);
                const runner = new run_1.Runner(nestedRunConfig);
                const resumeContextStrategy = resumeState?.contextStrategy ?? 'merge';
                const resumeContext = resumeContextStrategy === 'preferSerialized' ? undefined : runContext;
                let runInput = resolvedInput;
                if (details?.resumeState) {
                    if (resumeContextStrategy === 'preferSerialized' || !resumeContext) {
                        runInput = await runState_1.RunState.fromString(this, details.resumeState);
                    }
                    else {
                        if (resumeContextStrategy === 'merge' &&
                            context &&
                            resumeContext !== context) {
                            resumeContext._mergeApprovals(context.toJSON().approvals);
                        }
                        runInput = await runState_1.RunState.fromStringWithContext(this, details.resumeState, resumeContext, {
                            contextStrategy: resumeContextStrategy === 'replace' ? 'replace' : 'merge',
                        });
                    }
                }
                // Only flip to streaming mode when a handler is provided to avoid extra overhead for callers that do not need events.
                // Flip to streaming if either a legacy onStream callback or event handlers are registered; otherwise stay on the non-stream path to avoid extra overhead.
                const shouldStream = typeof onStream === 'function' || eventHandlers.size > 0;
                const configuredSignal = runOptions?.signal;
                const toolCallSignal = details?.signal;
                const { signal: combinedSignal, cleanup: cleanupSignalListeners } = configuredSignal && toolCallSignal
                    ? (0, abortSignals_1.combineAbortSignals)(configuredSignal, toolCallSignal)
                    : {
                        signal: toolCallSignal ?? configuredSignal,
                        cleanup: () => { },
                    };
                const runOptionsWithContext = {
                    ...(runOptions ?? {}),
                    context: runContext,
                    ...(combinedSignal ? { signal: combinedSignal } : {}),
                };
                try {
                    const result = shouldStream
                        ? await runner.run(this, runInput, {
                            ...runOptionsWithContext,
                            stream: true,
                        })
                        : await runner.run(this, runInput, {
                            ...runOptionsWithContext,
                        });
                    const streamPayload = {
                        agent: this,
                        toolCall: details?.toolCall,
                    };
                    if (shouldStream) {
                        // Cast through unknown: the async iterator shape matches and we want to drain the stream for side effects while keeping the public API stable.
                        const streamResult = result;
                        // Drain the stream to deliver every event to registered handlers; ensure completion awaited so the nested run finishes before returning.
                        for await (const event of streamResult) {
                            await emitEvent({
                                event,
                                ...streamPayload,
                            });
                        }
                        await streamResult.completed;
                    }
                    const completedResult = result;
                    if (completedResult.state instanceof runState_1.RunState) {
                        completedResult.state._agentToolInvocation = agentToolInvocation;
                    }
                    const completedResultWithAgentToolInvocation = completedResult;
                    const usesStopAtToolNames = typeof this.toolUseBehavior === 'object' &&
                        this.toolUseBehavior !== null &&
                        'stopAtToolNames' in this.toolUseBehavior;
                    if (typeof customOutputExtractor !== 'function' &&
                        usesStopAtToolNames) {
                        logger_1.default.debug(`You're passing the agent (name: ${this.name}) with toolUseBehavior.stopAtToolNames configured as a tool to a different agent; this may not work as you expect. You may want to have a wrapper function tool to consistently return the final output.`);
                    }
                    let outputText;
                    if (typeof customOutputExtractor === 'function') {
                        outputText = await customOutputExtractor(completedResultWithAgentToolInvocation);
                    }
                    else {
                        const finalOutputText = typeof completedResult.finalOutput !== 'undefined'
                            ? this.outputType === 'text'
                                ? String(completedResult.finalOutput)
                                : JSON.stringify(completedResult.finalOutput)
                            : undefined;
                        const rawResponses = completedResult.rawResponses;
                        const rawOutputText = rawResponses && rawResponses.length > 0
                            ? (0, messages_1.getOutputText)(rawResponses[rawResponses.length - 1])
                            : undefined;
                        const normalizedRawOutputText = typeof rawOutputText === 'string' && rawOutputText.trim() === ''
                            ? undefined
                            : rawOutputText;
                        const prefersFinalOutput = completedResult.state?._finalOutputSource === 'error_handler';
                        outputText = prefersFinalOutput
                            ? (finalOutputText ?? normalizedRawOutputText ?? '')
                            : (normalizedRawOutputText ?? finalOutputText ?? '');
                    }
                    if (details?.toolCall) {
                        (0, agentToolRunResults_1.saveAgentToolRunResult)(details.toolCall, completedResultWithAgentToolInvocation);
                    }
                    return outputText;
                }
                finally {
                    cleanupSignalListeners();
                }
            },
        });
        const agentTool = {
            ...baseTool,
            on: (name, handler) => {
                const set = eventHandlers.get(name) ?? new Set();
                set.add(handler);
                eventHandlers.set(name, set);
                return agentTool;
            },
        };
        (0, agentToolSourceRegistry_1.registerAgentToolSourceAgent)(agentTool, this);
        return agentTool;
    }
    /**
     * Returns the system prompt for the agent.
     *
     * If the agent has a function as its instructions, this function will be called with the
     * runContext and the agent instance.
     */
    async getSystemPrompt(runContext) {
        if (typeof this.instructions === 'function') {
            return await this.instructions(runContext, this);
        }
        return this.instructions;
    }
    /**
     * Returns the prompt template for the agent, if defined.
     *
     * If the agent has a function as its prompt, this function will be called with the
     * runContext and the agent instance.
     */
    async getPrompt(runContext) {
        if (typeof this.prompt === 'function') {
            return await this.prompt(runContext, this);
        }
        return this.prompt;
    }
    /**
     * Fetches the available tools from the MCP servers.
     * @returns the MCP powered tools
     */
    async getMcpTools(runContext) {
        if (this.mcpServers.length > 0) {
            return (0, mcp_1.getAllMcpTools)({
                mcpServers: this.mcpServers,
                runContext,
                agent: this,
                convertSchemasToStrict: false,
            });
        }
        return [];
    }
    /**
     * ALl agent tools, including the MCPl and function tools.
     *
     * @returns all configured tools
     */
    async getAllTools(runContext) {
        const mcpTools = await this.getMcpTools(runContext);
        const enabledTools = [];
        for (const candidate of this.tools) {
            if (candidate.type === 'function') {
                const maybeIsEnabled = candidate.isEnabled;
                const enabled = typeof maybeIsEnabled === 'function'
                    ? await maybeIsEnabled(runContext, this)
                    : typeof maybeIsEnabled === 'boolean'
                        ? maybeIsEnabled
                        : true;
                if (!enabled) {
                    continue;
                }
            }
            enabledTools.push(candidate);
        }
        return [...mcpTools, ...enabledTools];
    }
    hasExplicitToolConfig() {
        return this._toolsExplicitlyConfigured;
    }
    /**
     * Returns the handoffs that should be exposed to the model for the current run.
     *
     * Handoffs that provide an `isEnabled` function returning `false` are omitted.
     */
    async getEnabledHandoffs(runContext) {
        const handoffs = this.handoffs?.map((h) => (0, handoff_1.getHandoff)(h)) ?? [];
        const enabled = [];
        for (const handoff of handoffs) {
            if (await handoff.isEnabled({ runContext, agent: this })) {
                enabled.push(handoff);
            }
        }
        return enabled;
    }
    /**
     * Processes the final output of the agent.
     *
     * @param output - The output of the agent.
     * @returns The parsed out.
     */
    processFinalOutput(output) {
        if (this.outputType === 'text') {
            return output;
        }
        if (typeof this.outputType === 'object') {
            const parsed = JSON.parse(output);
            if ((0, typeGuards_1.isZodObject)(this.outputType)) {
                return this.outputType.parse(parsed);
            }
            return parsed;
        }
        throw new Error(`Unknown output type: ${this.outputType}`);
    }
    /**
     * Returns a JSON representation of the agent, which is serializable.
     *
     * @returns A JSON object containing the agent's name.
     */
    toJSON() {
        return {
            name: this.name,
        };
    }
}
exports.Agent = Agent;
