// @openai/agents-core/dist/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
exports.getLogger = getLogger;
const debug_1 = __importDefault(require("debug"));
const config_1 = require("./config.js");
/**
 * Get a logger for a given package.
 *
 * @param namespace - the namespace to use for the logger.
 * @returns A logger object with `debug` and `error` methods.
 */
function getLogger(namespace = 'openai-agents') {
    return {
        namespace,
        debug: (0, debug_1.default)(namespace),
        error: (...args) => console.error(...args),
        warn: (...args) => console.warn(...args),
        get dontLogModelData() {
            return config_1.logging.dontLogModelData;
        },
        get dontLogToolData() {
            return config_1.logging.dontLogToolData;
        },
    };
}
exports.logger = getLogger('openai-agents:core');
exports.default = exports.logger;
