// @openai/agents-core/dist/result.mjs
import { ReadableStream as _ReadableStream, TransformStream, Readable, } from '@openai/agents-core/_shims';
import { getTurnInput } from "./runner/items.mjs";
import logger from "./logger.mjs";
import { StreamEventTextStream } from "./types/protocol.mjs";
import { combineAbortSignalsWithOptions } from "./utils/abortSignals.mjs";
function createAbortHandlerRef() {
    const ref = {};
    const handler = () => {
        ref.current?.();
    };
    return { ref, handler };
}
class RunResultBase {
    state;
    constructor(state) {
        this.state = state;
    }
    /**
     * The history of the agent run. This includes the input items and the new items generated during
     * the agent run.
     *
     * This can be used as inputs for the next agent run.
     */
    get history() {
        return getTurnInput(this.input, this.newItems, this.state._reasoningItemIdPolicy);
    }
    /**
     * The new items generated during the agent run. These include things like new messages, tool
     * calls and their outputs, etc.
     *
     * It does not include information about the agents and instead represents the model data.
     *
     * For the output including the agents, use the `newItems` property.
     */
    get output() {
        return getTurnInput([], this.newItems, this.state._reasoningItemIdPolicy);
    }
    /**
     * A copy of the original input items.
     */
    get input() {
        return this.state._originalInput;
    }
    /**
     * The public run context for this run.
     */
    get runContext() {
        return this.state._context;
    }
    /**
     * Metadata about the nested `Agent.asTool()` invocation that produced this result, when applicable.
     */
    get agentToolInvocation() {
        return this.state._agentToolInvocation;
    }
    /**
     * The run items generated during the agent run. This associates the model data with the agents.
     *
     * For the model data that can be used as inputs for the next agent run, use the `output` property.
     */
    get newItems() {
        return this.state._generatedItems;
    }
    /**
     * The raw LLM responses generated by the model during the agent run.
     */
    get rawResponses() {
        return this.state._modelResponses;
    }
    /**
     * The last response ID generated by the model during the agent run.
     */
    get lastResponseId() {
        const responses = this.rawResponses;
        return responses && responses.length > 0
            ? responses[responses.length - 1].responseId
            : undefined;
    }
    /**
     * The last agent that was run
     */
    get lastAgent() {
        return this.state._currentAgent;
    }
    /**
     * The agent that should handle the next turn.
     * This is an alias for the last agent that completed a turn.
     */
    get activeAgent() {
        return this.lastAgent;
    }
    /**
     * Guardrail results for the input messages.
     */
    get inputGuardrailResults() {
        return this.state._inputGuardrailResults;
    }
    /**
     * Guardrail results for the final output of the agent.
     */
    get outputGuardrailResults() {
        return this.state._outputGuardrailResults;
    }
    /**
     * Guardrail results for tool inputs.
     */
    get toolInputGuardrailResults() {
        return this.state._toolInputGuardrailResults;
    }
    /**
     * Guardrail results for tool outputs.
     */
    get toolOutputGuardrailResults() {
        return this.state._toolOutputGuardrailResults;
    }
    /**
     * Any interruptions that occurred during the agent run for example for tool approvals.
     */
    get interruptions() {
        if (this.state._currentStep?.type === 'next_step_interruption') {
            return this.state._currentStep.data.interruptions;
        }
        return [];
    }
    /**
     * The final output of the agent. If the output type was set to anything other than `text`,
     * this will be parsed either as JSON or using the Zod schema you provided.
     */
    get finalOutput() {
        if (this.state._currentStep?.type === 'next_step_final_output') {
            return this.state._currentAgent.processFinalOutput(this.state._currentStep.output);
        }
        logger.warn('Accessed finalOutput before agent run is completed.');
        return undefined;
    }
}
/**
 * The result of an agent run.
 */
export class RunResult extends RunResultBase {
    constructor(state) {
        super(state);
    }
}
/**
 * The result of an agent run in streaming mode.
 */
export class StreamedRunResult extends RunResultBase {
    /**
     * The current agent that is running
     */
    get currentAgent() {
        return this.lastAgent;
    }
    /**
     * The current turn number
     */
    currentTurn = 0;
    /**
     * The maximum number of turns that can be run
     */
    maxTurns;
    #error = null;
    #combinedSignal;
    #abortSignalSnapshot;
    #abortController;
    #readableController;
    #readableStream;
    #completedPromise;
    #completedPromiseResolve;
    #completedPromiseReject;
    #cancelled = false;
    #streamLoopPromise;
    #abortHandler;
    #abortHandlerRef;
    #combinedSignalCleanup = () => { };
    constructor(result = {}) {
        super(result.state);
        this.#abortController = new AbortController();
        const { signal: combinedSignal, cleanup: cleanupCombinedSignal } = combineAbortSignalsWithOptions([result.signal, this.#abortController.signal], {
            onAbortSignalAnyError: (error) => {
                logger.debug(`AbortSignal.any failed, falling back: ${error}`);
            },
        });
        this.#combinedSignal = combinedSignal;
        this.#combinedSignalCleanup = cleanupCombinedSignal;
        this.#abortSignalSnapshot = combinedSignal;
        this.#readableStream = new _ReadableStream({
            start: (controller) => {
                this.#readableController = controller;
            },
            cancel: () => {
                if (!this.#abortController.signal.aborted) {
                    this.#abortController.abort();
                }
            },
        });
        this.#completedPromise = new Promise((resolve, reject) => {
            this.#completedPromiseResolve = resolve;
            this.#completedPromiseReject = reject;
        });
        if (this.#combinedSignal) {
            // Create the handler outside this scope so it cannot capture the run via lexical this.
            const { ref: abortRef, handler: handleAbort } = createAbortHandlerRef();
            abortRef.current = () => {
                this.#handleAbort();
            };
            this.#abortHandler = handleAbort;
            this.#abortHandlerRef = abortRef;
            if (this.#combinedSignal.aborted) {
                handleAbort();
            }
            else {
                this.#combinedSignal.addEventListener('abort', handleAbort, {
                    once: true,
                });
            }
        }
    }
    /**
     * @internal
     * Adds an item to the stream of output items
     */
    _addItem(item) {
        if (!this.cancelled) {
            this.#readableController?.enqueue(item);
        }
    }
    /**
     * @internal
     * Indicates that the stream has been completed
     */
    _done() {
        if (!this.cancelled && this.#readableController) {
            this.#readableController.close();
            this.#readableController = undefined;
            this.#completedPromiseResolve?.();
        }
        this.#detachAbortHandler();
    }
    /**
     * @internal
     * Handles an error in the stream loop.
     */
    _raiseError(err) {
        if (!this.cancelled && this.#readableController) {
            this.#readableController.error(err);
            this.#readableController = undefined;
        }
        this.#error = err;
        this.#completedPromiseReject?.(err);
        this.#completedPromise.catch((e) => {
            logger.debug(`Resulted in an error: ${e}`);
        });
        this.#detachAbortHandler();
    }
    /**
     * Returns true if the stream has been cancelled.
     */
    get cancelled() {
        return this.#cancelled;
    }
    /**
     * Returns the underlying readable stream.
     * @returns A readable stream of the agent run.
     */
    toStream() {
        return this.#readableStream;
    }
    /**
     * Await this promise to ensure that the stream has been completed if you are not consuming the
     * stream directly.
     */
    get completed() {
        return this.#completedPromise;
    }
    /**
     * Error thrown during the run, if any.
     */
    get error() {
        return this.#error;
    }
    toTextStream(options = {}) {
        const stream = this.#readableStream.pipeThrough(new TransformStream({
            transform(event, controller) {
                if (event.type === 'raw_model_stream_event' &&
                    event.data.type === 'output_text_delta') {
                    const item = StreamEventTextStream.parse(event.data);
                    controller.enqueue(item.delta);
                }
            },
        }));
        if (options.compatibleWithNodeStreams) {
            return Readable.fromWeb(stream);
        }
        return stream;
    }
    [Symbol.asyncIterator]() {
        return this.#readableStream[Symbol.asyncIterator]();
    }
    /**
     * @internal
     * Sets the stream loop promise that completes when the internal stream loop finishes.
     * This is used to defer trace end until all agent work is complete.
     */
    _setStreamLoopPromise(promise) {
        this.#streamLoopPromise = promise;
    }
    /**
     * @internal
     * Returns a promise that resolves when the stream loop completes.
     * This is used by the tracing system to wait for all agent work before ending the trace.
     */
    _getStreamLoopPromise() {
        return this.#streamLoopPromise;
    }
    /**
     * @internal
     * Returns the abort signal that should be used to cancel the streaming run.
     */
    _getAbortSignal() {
        return this.#abortSignalSnapshot ?? this.#combinedSignal;
    }
    #handleAbort() {
        if (this.#cancelled) {
            this.#detachAbortHandler();
            return;
        }
        this.#cancelled = true;
        const controller = this.#readableController;
        this.#readableController = undefined;
        if (controller) {
            try {
                controller.close();
            }
            catch (err) {
                logger.debug(`Failed to close readable stream on abort: ${err}`);
            }
        }
        this.#completedPromiseResolve?.();
        this.#detachAbortHandler();
    }
    #detachAbortHandler() {
        // Clear the indirection first so a retained signal cannot keep this run alive.
        if (this.#abortHandlerRef) {
            this.#abortHandlerRef.current = undefined;
        }
        if (this.#combinedSignal && this.#abortHandler) {
            try {
                this.#combinedSignal.removeEventListener('abort', this.#abortHandler);
            }
            catch (err) {
                logger.debug(`Failed to remove abort listener: ${err}`);
            }
        }
        try {
            this.#combinedSignalCleanup();
        }
        catch (err) {
            logger.debug(`Failed to clean up combined abort listeners: ${err}`);
        }
        this.#combinedSignalCleanup = () => { };
        this.#abortHandler = undefined;
        this.#abortHandlerRef = undefined;
    }
}
