// @openai/agents-core/dist/toolIdentity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FUNCTION_TOOL_NAMESPACE_DESCRIPTION = exports.FUNCTION_TOOL_NAMESPACE = exports.toolQualifiedName = exports.toolDisplayName = void 0;
exports.getToolCallNamespace = getToolCallNamespace;
exports.getToolCallName = getToolCallName;
exports.getToolCallQualifiedName = getToolCallQualifiedName;
exports.getToolCallDisplayName = getToolCallDisplayName;
exports.resolveFunctionToolCallName = resolveFunctionToolCallName;
exports.getExplicitFunctionToolNamespace = getExplicitFunctionToolNamespace;
exports.getFunctionToolNamespace = getFunctionToolNamespace;
exports.getFunctionToolNamespaceDescription = getFunctionToolNamespaceDescription;
exports.getFunctionToolQualifiedName = getFunctionToolQualifiedName;
exports.getFunctionToolDisplayName = getFunctionToolDisplayName;
exports.matchesFunctionToolName = matchesFunctionToolName;
const tooling_1 = require("./tooling.js");
var tooling_2 = require("./tooling.js");
Object.defineProperty(exports, "toolDisplayName", { enumerable: true, get: function () { return tooling_2.toolDisplayName; } });
Object.defineProperty(exports, "toolQualifiedName", { enumerable: true, get: function () { return tooling_2.toolQualifiedName; } });
exports.FUNCTION_TOOL_NAMESPACE = Symbol('functionToolNamespace');
exports.FUNCTION_TOOL_NAMESPACE_DESCRIPTION = Symbol('functionToolNamespaceDescription');
function isNonEmptyString(value) {
    return typeof value === 'string' && value.length > 0;
}
function getToolCallNamespace(toolCall) {
    return isNonEmptyString(toolCall.namespace) ? toolCall.namespace : undefined;
}
function getToolCallName(toolCall) {
    return isNonEmptyString(toolCall.name) ? toolCall.name : undefined;
}
function getToolCallQualifiedName(toolCall) {
    return (0, tooling_1.toolQualifiedName)(getToolCallName(toolCall), getToolCallNamespace(toolCall));
}
function getToolCallDisplayName(toolCall) {
    return (0, tooling_1.toolDisplayName)(getToolCallName(toolCall), getToolCallNamespace(toolCall));
}
function isTopLevelDeferredFunctionTool(candidate, bareName) {
    const tool = candidate;
    return (tool?.type === 'function' &&
        tool.name === bareName &&
        tool.deferLoading === true &&
        !isNonEmptyString(tool?.[exports.FUNCTION_TOOL_NAMESPACE]));
}
function resolveFunctionToolCallName(toolCall, availableToolNames) {
    const bareName = getToolCallName(toolCall);
    const namespace = getToolCallNamespace(toolCall);
    if (bareName && !namespace && availableToolNames.has(bareName)) {
        return bareName;
    }
    const qualifiedName = getToolCallQualifiedName(toolCall);
    const bareTool = bareName && typeof availableToolNames.get === 'function'
        ? availableToolNames.get(bareName)
        : undefined;
    const preferBareSelfNamespacedDeferredTool = bareName &&
        namespace === bareName &&
        availableToolNames.has(bareName) &&
        isTopLevelDeferredFunctionTool(bareTool, bareName);
    if (qualifiedName && availableToolNames.has(qualifiedName)) {
        if (preferBareSelfNamespacedDeferredTool) {
            return bareName;
        }
        return qualifiedName;
    }
    if (bareName && namespace === bareName && availableToolNames.has(bareName)) {
        return bareName;
    }
    return qualifiedName ?? bareName;
}
function getExplicitFunctionToolNamespace(tool) {
    const candidate = tool;
    return isNonEmptyString(candidate?.[exports.FUNCTION_TOOL_NAMESPACE])
        ? candidate[exports.FUNCTION_TOOL_NAMESPACE]
        : undefined;
}
function getFunctionToolNamespace(tool) {
    return getExplicitFunctionToolNamespace(tool);
}
function getFunctionToolNamespaceDescription(tool) {
    const candidate = tool;
    return isNonEmptyString(candidate?.[exports.FUNCTION_TOOL_NAMESPACE_DESCRIPTION])
        ? candidate[exports.FUNCTION_TOOL_NAMESPACE_DESCRIPTION]
        : undefined;
}
function getFunctionToolQualifiedName(tool) {
    const candidate = tool;
    return (0, tooling_1.toolQualifiedName)(isNonEmptyString(candidate?.name) ? candidate.name : undefined, getFunctionToolNamespace(tool));
}
function getFunctionToolDisplayName(tool) {
    const candidate = tool;
    return (0, tooling_1.toolDisplayName)(isNonEmptyString(candidate?.name) ? candidate.name : undefined, getFunctionToolNamespace(tool));
}
function matchesFunctionToolName(tool, candidate) {
    if (!isNonEmptyString(candidate)) {
        return false;
    }
    const bareName = getFunctionToolName(tool);
    if (bareName === candidate) {
        return true;
    }
    return getFunctionToolQualifiedName(tool) === candidate;
}
function getFunctionToolName(tool) {
    const candidate = tool;
    return isNonEmptyString(candidate?.name) ? candidate.name : undefined;
}
